### ISSUE TITLE
[MU4 Issue] Incorrect repeat when section break is present

### ISSUE THREAD

**Issue by SamYeh** [musescore/MuseScore#14262]
**Describe the bug**
When a measure has both a right repeat sign and a section break, followed by more or measures, the pause for the section break is included in the repeat.

**To Reproduce**
![image](https://user-images.githubusercontent.com/10248807/199401478-933679da-a2fe-494b-8a2c-b635901c4ef4.png)
Play and observe that there is a 3-second pause after the first two bars have been played once.

**Expected behavior**
The first two bars should be played twice without pause in between.

**Platform information**
 - OS: Windows 11

**Comment by DaddyLudwig** [comment]
I’m pretty sure this has always been the case. I myself have to do one of 2 things for this, those being:

1. Set pause to 0 for section break(like if I have repeats and attaca)
2. Place the section break one measure after the repeat and set the measure invisible in measure properties(like in the development and recapitulation repeat of a Mozart Sonata)

**Comment by SamYeh** [comment]
But is that the expected behavior, though?

**Comment by DaddyLudwig** [comment]
I’d say yes because it was like this in 3.6 and I’ve never seen talk about changing the behavior of the section break on repeats, just the 2 workarounds for it I mentioned.

**Comment by MarcSabatella** [comment]
Definitely not *intended*, but yes, it's a bug in 3.6.2 as well.  There have actually been quite a few reports of this and some attempts made at coming up with fixes, but no success yet.

**Comment by jeetee** [comment]
As mentioned in https://musescore.org/en/node/32696 I'm willing to try and take this up again now that MPE is starting to settle down. The RepeatList logic already contains some preparation information on when the pause should be applied, but we still need to look into how to tie that information in with the code that generates the playback information (which in MS3 also meant that first the TempoMap needed to be made repeat-aware).

**Comment by BanjoJake** [comment]
Also happens using loop playback without repeat bars, if measure ends in a rest.  
[Playback Loop Bugs when Bar ends on Rest.zip](https://github.com/musescore/MuseScore/files/12413085/Playback.Loop.Bugs.when.Bar.ends.on.Rest.zip)

**Comment by adelara** [comment]
Happy Friday, March/1/2024... that bug is still nagging isn't ? I'd think 10 years is a pretty good period to have it figured out... luckily I dont ran into it often, but today, I did - https://musescore.org/en/node/32696

**Comment by cbjeukendrup** [comment]
Perhaps the [newly added](https://github.com/musescore/MuseScore/commit/b2f2205c27e4d7f35eb877c656c25cb0bb04d370) `TempoMap::pauseSecs` can help to fix this issue.

**Comment by Jojo-Schmitz** [comment]
@jeetee ?

**Comment by henkdegroot** [comment]
@jeetee @Jojo-Schmitz would this issue also cover this bug when using a Da Capo and a Section Break? Or should a separate issue be reported for that?

**Comment by jeetee** [comment]
`TempoMap::pauseSecs` in itself doesn't offer the solution.
The root problem is that the tempomap (which contains those pauses) doesn't have an unrolled version and as such doesn't differentiate on which repeat/jump you collect its information.

Implementing an unrolled tempomap to match the unrolled score would address that and also allow for correctly choosing the correct repeat/jump when calculating back from a time to a position (such as when scrubbing the playback thumb on the playback time UI)

@henkdegroot This indeed should also solve interactions between jumps and the section break. Of course we'll need a double check (because I wrote the foundation for this a few years ago by now..), but https://github.com/musescore/MuseScore/blob/master/src/engraving/dom/repeatlist.cpp#L1034-L1045 should ensure that the pause property of the repeatSegment is only set when the section break is encountered at the end of the section (and thus not upon a repeat/jump, which is already processed higher up in that function).

**Comment by Jojo-Schmitz** [comment]
Came up once again in https://musescore.org/en/node/368534

**Comment by PaulJoosten** [comment]
Section ends serve two purposes: in layout (to not show key and time signatures from the next movement) and playback (to add a pause before moving to the next movement. I need both to coincide. When the section ends in a repeat, the pause is inserted already when the repeat is played and it gets even worse when the section ends in a da capo al fine. In the pictures you see what I want it to look like in the first, but it only plays correct (but without a break...) in the second.

<img width="436" height="554" alt="Image" src="https://github.com/user-attachments/assets/b870dc2e-bc40-4598-aaea-b28e8f8985ca" />
<img width="406" height="595" alt="Image" src="https://github.com/user-attachments/assets/0665358d-83d2-4bd3-a4db-d9f22c3ec3b8" />

Apparently this bug has been known for quite a long time. If this takes much longer to fix, is there some workaround where both the layout and the playback work like they should?

**Comment by DaddyLudwig** [comment]
> Apparently this bug has been known for quite a long time. If this takes much longer to fix, is there some workaround where both the layout and the playback work like they should?

What I've done is I'll add the section break on the measure right after the repeat rather than on the repeat, and then I'll select that measure I just added the section break to, right click and go to Measure Properties, and then hide that measure on all staves. That way I get a visual separation between movements without getting the pause on the repeat.

In the case of a minuet da capo, this gets more complicated. The initial step of putting the section break on the measure after the repeat is still the same, but now, instead of a fine at the end of the minuet section, I need to put a To Coda and make it look like "Fine" and then add a coda sign to that measure that I just added the section break to, and make that invisible _before_ I hide the extra measure. That way, I still get a pause, even if I have another movement after the minuet (which I often do with me inputting classical music pieces by Mozart and the like into MuseScore)

**Comment by PaulJoosten** [comment]
Thanks! I tried all that but even when hidden the extra measure still takes up space altering the desired layout. Looks like I still need to choose between the correct layout and the breaks or do I miss something?

**Comment by Jojo-Schmitz** [comment]
Came up again in https://musescore.org/en/node/382535

**Comment by PaulJoosten** [comment]
What if we think a bit out of the box and change the pause to act at the beginning of the new section? Might be easier to program and would solve all the issues.

**Comment by Jojo-Schmitz** [comment]
That may not fix it, like when that new sections **starts** with a repeat, so this would just shift the problem from end repeat to start repeat. The pause needs to be just in between (and there only)

**Comment by PaulJoosten** [comment]
I'll gladly leave that to the programmers to find out. I've only recently upgraded to 4, in 3 I liked the feature when you use the metronome it can give you a measure in advance, that seems to have got lost too?

**Comment by Jojo-Schmitz** [comment]
That'll be back in 4.6

**Comment by Jojo-Schmitz** [comment]
Came up again in https://musescore.org/en/node/389210

**Comment by Jojo-Schmitz** [comment]
Please reopen this issue: it is not resolved.
Check the attached score so see it failing.

[repeats@section_breaks.zip](https://github.com/user-attachments/files/31793375/repeats%40section_breaks.zip)

See https://github.com/Jojo-Schmitz/MuseScore/pull/1276 for a fully working patch for my  3.7
And #34796 for a (hopefully, testing needed) better fix for 4.7 (will port it to main too, if the testing is successful)
Edit: Unfortunately it doesn't work (yet?)

**Comment by DmitryArefiev** [comment]
@Jojo-Schmitz The initial case is fixed. Your case is different. And as I understand, the section break is not designed for using it in the middle of piece/section (https://handbook.musescore.org/formatting/using-sections-for-multiple-movements-or-songs#section-break-properties) like in your case:

<img width="1824" height="1272" alt="Image" src="https://github.com/user-attachments/assets/c6c1cabe-f836-4aa0-925b-64543ea624a9" />

**Comment by Jojo-Schmitz** [comment]
What? Section breaks are exactly about that: starting a new piece within a score. 
And if the piece ending in a section break also ends with an end repeat or a jump, at the repeat the pause intended for between the pieces should not happen, but only after that piece really finished, i.e. before the next piece starts.

That section break on the frame is a fluke, and shouldn't be there (my mistake), but doesn't make a difference. The next one is working just fine (it pauses as it should), the ones after don't, they do pause even for the end repeats, D.C., etc. (and additionaly, rightly, before the next piece starts)

The (initial) problem had been fixed (or rather been [duct-taped](https://github.com/musescore/MuseScore/pull/34350#issuecomment-5333850117)), but it is broken again.

**Comment by DmitryArefiev** [comment]
> What? Section breaks are exactly about that: starting a new piece within a score.

Yes.. I mean that the section break is a separator between **two pieces** in **one score** (like a final barline) and it's the starting point for next repeat barline. In your case I didn't see a final barline (or double barline) in section break points, that's why I didn't understand what exactly you mean (where the problem is)

The **initial** case still looks fine in **main**:

https://github.com/user-attachments/assets/e1fa291f-e3aa-40cb-b637-e944d9b1a1c9

However, your case is different, your score have multiple section breaks and yes, it **doesn't work** in **main**

https://github.com/user-attachments/assets/cf900eb8-449e-4733-8770-1c9cb5523917

But it **does work** in **4.7.5**:

https://github.com/user-attachments/assets/dc9c9814-b9a9-43f5-9b82-a925da9f1080


Please create a **separate issue** (and link to your [PR](https://github.com/musescore/MuseScore/pull/34796))

**Comment by Jojo-Schmitz** [comment]
See #34870 and #34871

**Comment by DmitryArefiev** [comment]
Thanks!
